runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -616,25 +633,62 @@ private void syncStreamUnsafe(ISMRStream stream, long timestamp) {
-                        } else if (pendingUpcalls.contains(entry.getGlobalAddress())) {
-                            log.debug("Sync[{}] Upcall Result {}",
-                                    this, entry.getGlobalAddress());
-                            upcallResults.put(entry.getGlobalAddress(), res == null
-                                    ? NullValue.NULL_VALUE : res);
-                            pendingUpcalls.remove(entry.getGlobalAddress());
+                        Object res = applyUpdateUnsafe(record);
+
+                        if (fromWriteSMRStream) {
+                            record.setUpcallResult(res);
+                        } else { // materialized log
+                            if (pendingUpcalls.contains(record.getGlobalAddress())) {
+                                log.debug("Sync[{}] Upcall Result {}",
+                                        this, record.getGlobalAddress());
+                                upcallResults.put(record.getGlobalAddress(), res == null
+                                        ? NullValue.NULL_VALUE : res);
+                                pendingUpcalls.remove(record.getGlobalAddress());
+                            }
+
+                            // Record with global address larger than lastSyncAddress indicates that all SMRRecords
+                            // from last global address has been applied. Therefore, syncTail could be updated.
+                            if (lastSyncAddress.get() != Address.NON_ADDRESS &&
+                                    record.getGlobalAddress() > lastSyncAddress.get()) {
+                                syncTail = lastSyncAddress.get();
+                            }
+
+                            // to check whether the SMRRecord is first encountered by VLO
+                            if (record.getGlobalAddress() > syncTail) {
+                                IGarbageIdentificationFunction garbageIdentifyFunction =
+                                        garbageIdentifyFunctionMap.get(record.getSMRMethod());
+                                if (garbageIdentifyFunction != null) {
+                                    List<Object> garbage =
+                                            garbageIdentifyFunction.identify(record.locator, record.getSMRArguments());
+                                    // TODO(xin): handle garbage
+                                }
+
+                                lastSyncAddress.set(record.getGlobalAddress());
+                            }
                         }
-                        entry.setUpcallResult(res);
+
+                        record.setUpcallResult(res);
                     } catch (Exception e) {
                         log.error("Sync[{}] Error: Couldn't execute upcall due to {}", this, e);
                         throw new UnrecoverableCorfuError(e);
                     }
                 });
+
+
+        if (!fromWriteSMRStream) {
+            syncTail = lastSyncAddress.get();
```

Comment:
This could regress syncTail, you can use Math.max()

Reply:
done.